### PR TITLE
[lldb][libc++] Adds system_clock data formatters.

### PR DESCRIPTION
**Pull request by mordante**

### CODE REVIEW

File: lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp [hunk 1]
```diff
@@ -1073,18 +1073,79 @@ bool lldb_private::formatters::LibcxxWStringViewSummaryProvider(
...
+
+  const int days = ptr_sp->GetValueAsSigned(0);
+  if (days < chrono_timestamp_min || days > chrono_timestamp_max)
+    stream.Printf("timestamp=%d days", days);
+
+  else {
+    const std::time_t seconds = std::time_t(86400) * days;
+
+    std::array<char, 128> str;
+    std::strftime(str.data(), str.size(), "%FZ", gmtime(&seconds));
```
Comment (Michael137) [inline lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp:1142]: is it worth checking the return of `strftime` here?

File: lldb/test/API/functionalities/data-formatter/data-formatter-stl/libcxx/chrono/TestDataFormatterLibcxxChrono.py [hunk 2]
```diff
@@ -32,6 +32,106 @@ def test_with_run_command(self):
...
+        self.expect(
+            "frame variable sd_min",
+            substrs=["sd_min = timestamp=-2147483648 days"],
+        )
+        self.expect(
+            "frame variable sd_max",
+            substrs=["sd_max = timestamp=2147483647 days"],
+        )
+
+        # self.expect("frame variable sd_min", substrs=["sd_min = date=1970-01-01Z timestamp=0 days"])
```
Comment (Michael137) [inline lldb/test/API/functionalities/data-formatter/data-formatter-stl/libcxx/chrono/TestDataFormatterLibcxxChrono.py:131]: Why are these commented?

**Review by Michael137 — APPROVED** [review]
LGTM (with minor questions)

File: lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp (hunk 1 above)
Comment (mordante) [reply lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp:1142]: I'll them, but they should never trigger.

File: lldb/test/API/functionalities/data-formatter/data-formatter-stl/libcxx/chrono/TestDataFormatterLibcxxChrono.py (hunk 2 above)
Comment (mordante) [reply lldb/test/API/functionalities/data-formatter/data-formatter-stl/libcxx/chrono/TestDataFormatterLibcxxChrono.py:131]: nice catch, leftover copy-pasted

**Comment by jasonmolenda** [comment]
@mordante I'm seeing failures on the macOS bots (and on my desktop) with TestDataFormatterLibcxxChrono.py, the test

        self.expect(
            "frame variable ss_neg_date_time",
            substrs=[
                "ss_neg_date_time = date/time=-32767-01-01T00:00:00Z timestamp=-1096193779200 s"
            ],
        )

but we're getting an unsigned timestamp from lldb instead:

```
(lldb) fr v ss_neg_date_time
(std::chrono::sys_seconds) ss_neg_date_time = date/time=-32767-01-01T00:00:00Z timestamp=18446742977515772416 s
```

so the test fails, from the `%ld` formater in formatters::LibcxxChronoSysSecondsSummaryProvider I suppose.  Both of these are formattings of the value 0xffffff00c5c25a00, I'm honestly not sure why it's formatted as an unsigned value on macOS for me, I'm guessing it isn't on linux or some platform?

**Comment by jasonmolenda** [comment]
I commented out the two tests that are failing on macOS temporarily so you'd have a chance to look at this.  If you don't have access to macOS (I assume this is working on Linux or whatever), @Michael137 or I can look into it.  Thanks.  

```
commit ba45ad160e3f329aeb02c19eaf18af27fa423d85
Author: Jason Molenda <jmolenda@apple.com>
Date:   Thu Jan 25 16:30:14 2024 -0800

    Temporarily disable two libcxx chrono formatter tests
```

**Comment by mordante** [comment]
@jasonmolenda I'm not using a mac, but based on your results I expect that `std::size_t` is a 64-bit _unsigned_ integer type. Is that correct?

**Comment by Michael137** [comment]
tbf, we are casting the value to an uint64_t and printing it with a `PRIu64`, right?:
```
stream.Printf("timestamp=%" PRIu64 " s", static_cast<uint64_t>(seconds))
```

So how would we ever expect a `-1096193779200` here?

**Comment by jasonmolenda** [comment]
Agreed, if I change this to 

```
diff --git a/lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp b/lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp
index d0bdbe1fd4d..b37544f6dd3 100644
--- a/lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp
+++ b/lldb/source/Plugins/Language/CPlusPlus/LibCxx.cpp
@@ -1105,7 +1105,7 @@ bool lldb_private::formatters::LibcxxChronoSysSecondsSummaryProvider(
 
   const std::time_t seconds = ptr_sp->GetValueAsSigned(0);
   if (seconds < chrono_timestamp_min || seconds > chrono_timestamp_max)
-    stream.Printf("timestamp=%" PRIu64 " s", static_cast<uint64_t>(seconds));
+    stream.Printf("timestamp=%" PRId64 " s", static_cast<int64_t>(seconds));
   else {
     std::array<char, 128> str;
     std::size_t size =
@@ -1113,8 +1113,8 @@ bool lldb_private::formatters::LibcxxChronoSysSecondsSummaryProvider(
     if (size == 0)
       return false;
 
-    stream.Printf("date/time=%s timestamp=%" PRIu64 " s", str.data(),
-                  static_cast<uint64_t>(seconds));
+    stream.Printf("date/time=%s timestamp=%" PRId64 " s", str.data(),
+                  static_cast<int64_t>(seconds));
   }
 
   return true;
```

the tests pass on Darwin.  On Darwin `time_t` is `long`, a signed value - and given that this can represent negative times, that makes sense.  The test is checking that PRIu64 prints a negative value but we're casting this time_t to a uint64_t, I don't understand why this would print as a negative value on Linux (and pass the tests).

**Comment by jasonmolenda** [comment]
Created a new PR with my proposed fix https://github.com/llvm/llvm-project/pull/81078

**Pull request merged**